parrot.rs

```diff
@@ -1,73 +1,62 @@
-use core::sync::atomic::{AtomicUsize, Ordering};
 use kernel::{
     bindings::msleep,
-    chrdev::Registration,
-    file::{File, Operations, ToUse},
+    declare_file_operations,
+    file::{File, Operations},
     io_buffer::IoBufferWriter,
+    miscdev::Registration,
     prelude::*,
 };
 
 mod frames;
-use frames::*;
+use frames::{calc_frame_and_offset, FRAMES};
 
 module! {
     type: ParrotSafe,
     name: b"party_parrot",
-    author: b"John Baublitz (john.m.baublitz@gmail.com)",
+    author: b"John Baublitz <john.m.baublitz@gmail.com>",
     description: b"Get the party started",
     license: b"GPL",
 }
 
-const FRAMES: [&str; 10] = [
-    FRAME0, FRAME1, FRAME2, FRAME3, FRAME4, FRAME5, FRAME6, FRAME7, FRAME8, FRAME9,
-];
-
-struct ParrotSafe(Pin<Box<Registration<1>>>);
+struct ParrotSafe(Pin<Box<Registration<ParrotOps>>>);
 
 impl kernel::Module for ParrotSafe {
-    fn init(name: &'static CStr, module: &'static ThisModule) -> Result<Self> {
-        let mut registration = Registration::new_pinned(name, 0, module)?;
-        registration.as_mut().register::<ParrotOps>()?;
-        Ok(ParrotSafe(registration))
+    fn init(_: &'static CStr, _: &'static ThisModule) -> Result<Self> {
+        Ok(ParrotSafe(Registration::new_pinned(
+            format_args!("parrot"),
```

Comment:
Not that it matters much, but we have a shorter macro for this declared in the prelude: `fmt`